compiler/vm/vm.nim

```diff
@@ -653,73 +828,90 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
+      ensureAtomKind(akFloat)
       regs[ra].floatVal = cast[float64](regs[rb].intVal)
 
     of opcCastPtrToInt: # RENAME opcCastPtrOrRefToInt
-      decodeBImm(rkInt)
+      decodeBImm(akInt)
+      var intVal : int
       case imm
       of 1: # PtrLikeKinds
-        case regs[rb].kind
-        of rkNode:
-          regs[ra].intVal = cast[int](regs[rb].node.intVal)
-        of rkNodeAddr:
-          regs[ra].intVal = cast[int](regs[rb].nodeAddr)
-        else:
+        case regs[rb].atomVal.kind
+        of akPtr, akRef:
+          intVal = cast[int](regs[rb].atomVal.ptrVal)
+        of akPointer:
+          intVal = cast[int](regs[rb].atomVal.pointerVal)
+        #of rkNodeAddr:
+        #  regs[ra].intVal = cast[int](regs[rb].nodeAddr)
+        of akInt, akFloat, akArray, akSeq, akString:
           stackTrace(
             c, tos, pc,
             reportStr(
-              rsemVmErrInternal, "opcCastPtrToInt: got " & $regs[rb].kind))
+              rsemVmErrInternal, "opcCastPtrToInt: got " & $regs[rb].atomVal.kind))
+          exitWithError()
 
       of 2: # tyRef
-        regs[ra].intVal = cast[int](regs[rb].node)
-      else: assert false, $imm
+        intVal = cast[int](regs[rb].atomVal.ptrVal)
+      else: assert false, $imm # TODO: should this assertion really be debug-mode only?
+      regs[ra].atomVal.intVal = intVal
     of opcCastIntToPtr:
       let rb = instr.regB
-      let typ = regs[ra].node.typ
-      let node2 = newNodeIT(nkIntLit, c.debug[pc], typ)
-      case regs[rb].kind
-      of rkInt: node2.intVal = regs[rb].intVal
-      of rkNode:
+      let typ = regs[ra].atomVal.ptrTyp
+      var ptrVal : pointer
+      case regs[rb].atomVal.kind
+      of akInt:
+        ptrVal = cast[pointer](regs[rb].intVal)
+      #of rkInt: ptrVal = regs[rb].intVal
+      #[of rkNode:
+        # XXX should this case even be supported?
+        # XXX Why require b to be pointer like if we're casting from an int? 
+        #[
         if regs[rb].node.typ.kind notin PtrLikeKinds:
           stackTrace(c, tos, pc, reportStr(
             rsemVmErrInternal,
             "opcCastIntToPtr: regs[rb].node.typ: " & $regs[rb].node.typ.kind))
+        ]#
 
-        node2.intVal = regs[rb].node.intVal
-
+        #ptrVal = regs[rb].node
+        {.warning: "missing".}
+        doAssert false, "?"
+      ]#
       else:
         stackTrace(c, tos, pc, reportStr(
           rsemVmErrInternal,
-          "opcCastIntToPtr: regs[rb].kind: " & $regs[rb].kind))
+          "opcCastIntToPtr: regs[rb].atomVal.kind: " & $regs[rb].atomVal.kind))
+        exitWithError()
 
-      regs[ra].node = node2
+      regs[ra].atomVal.ptrVal = ptrVal
     of opcAsgnComplex:
       asgnComplex(regs[ra], regs[instr.regB])
     of opcFastAsgnComplex:
       fastAsgnComplex(regs[ra], regs[instr.regB])
     of opcAsgnRef:
+      {.warning: "Look at this case again".}
       asgnRef(regs[ra], regs[instr.regB])
     of opcNodeToReg:
+      # XXX why is this needed? For what?
```

Comment:
Hints on vmgen, most likely.

Reply:
This is used for `mReset` (`system.reset()`) implementation it seems: https://github.com/nim-works/nimskull/blob/9d0baf777824711d9ef5d467a3ad0e7b9234b054/compiler/vm/vmgen.nim#L1194-L1199